packages/snfoundry/contracts/src/test/TestContract.cairo

```diff
@@ -227,576 +373,375 @@ fn test_join_session() {
+    let ver_rand_num1 = VerifiableRandomNumber { random_number: 6, v: 1, r: 1, s: 1 };
+    let ver_rand_num2 = VerifiableRandomNumber { random_number: 1, v: 1, r: 1, s: 1 };
     let ver_rand_num_array = array![ver_rand_num0, ver_rand_num1, ver_rand_num2];
-    ludo_dispatcher.play(session_id, ludo_move, ver_rand_num_array);
-    let (session_data, ludo_session_status) = ludo_dispatcher.get_session_status(session_id);
-    println!("{:?}", session_data);
-    let (user0, user1, _, _) = ludo_session_status.users;
-    let (user0_pin_0_pos, _, _, _) = user0.player_tokens_position;
-    let (user1_pin_0_pos, _, _, _) = user1.player_tokens_position;
+    let (user0, user1, _, _) = player_move(context, @ludo_move, player_0, ver_rand_num_array);
     let new_expected_user0_pin_0_pos = expected_user0_pin_0_pos + 13;
-    println!("-- User 0 pin 0 pos: {:?}", user0_pin_0_pos);
-    println!("-- User 1 pin 0 pos: {:?}", user1_pin_0_pos);
-    assert_eq!(user0_pin_0_pos, new_expected_user0_pin_0_pos);
-    assert_eq!(user1_pin_0_pos, 0);
+    assert_position_0_eq(@user0, new_expected_user0_pin_0_pos);
+    assert_position_0_eq(@user1, 0);
 }
 #[test]
 fn test_user0_pin0_wins() {
-    let ludo_contract = deploy_ludo_contract();
-    let ludo_dispatcher = ILudoDispatcher { contract_address: ludo_contract };
-    let marquis_game_dispatcher = IMarquisGameDispatcher { contract_address: ludo_contract };
-    let token = ZERO_TOKEN();
-    let amount = 0;
     let player_0 = PLAYER_0();
-    cheat_caller_address(ludo_contract, player_0, CheatSpan::TargetCalls(1));
-    let session_id = marquis_game_dispatcher.create_session(token, amount);
     let player_1 = PLAYER_1();
-    cheat_caller_address(ludo_contract, player_1, CheatSpan::TargetCalls(1));
-    marquis_game_dispatcher.join_session(session_id);
     let player_2 = PLAYER_2();
-    cheat_caller_address(ludo_contract, player_2, CheatSpan::TargetCalls(1));
-    marquis_game_dispatcher.join_session(session_id);
     let player_3 = PLAYER_3();
-    cheat_caller_address(ludo_contract, player_3, CheatSpan::TargetCalls(1));
-    marquis_game_dispatcher.join_session(session_id);
-    let (session_data, ludo_session_status) = ludo_dispatcher.get_session_status(session_id);
-    let status = session_data.status;
-    let expected_status = 2; // can play
-    assert_eq!(status, expected_status);
-    println!("{:?}", session_data);
-    println!("{:?}", ludo_session_status);
 
-    let ludo_move = LudoMove { token_id: 0 };
-    let ludo_move1 = ludo_move.clone();
-    let ludo_move2 = ludo_move.clone();
-    let ludo_move3 = ludo_move.clone();
-    let ver_rand_num0 = VerifiableRandomNumber { random_number: 6, v: 1, r: 1, s: 1, };
+    let (context, _) = setup_game_4_players(ZERO_TOKEN(), 0);
+
+    let ver_rand_num0 = VerifiableRandomNumber { random_number: 6, v: 1, r: 1, s: 1 };
     // can actually move from here
     let ver_rand_num1 = ver_rand_num0.clone();
```

Comment:
can we use some loop here?

Reply:
I don't touch "move datasets" because it's tricky to find the balance between readability and over engineering.
But maybe we can find an elegante solution